### ISSUE TITLE
Attention Behavior Test Not Working On Cuda

### ISSUE THREAD

**Issue by Eupham** [Eupham/KernelDev#179]
I ran this on an h100 gpu and still got failures """⚠ Kernel execution failed as expected on CPU (requires CUDA)""".  Also I don't feel like the resulting masks communicate the intended behavior.  It doesn't demonstrate the isolation of the spans, or that they can view the context, it doesn't demonstrate the causal behavior of the context

For example the expected behaviors are as follows
any token before cls and including cls are bidirectional for both tasks, pad tokens are ignored
for teacher forcing all tokens after cls should be causal and should see the cls token
for cocktail party there are 4 parts.

the prefix up to cls which we've already described
the context which should behave causally and may include the [mask] token (no special behavior for mask)
3 after the context are the span islands which are [span]candidate text[es] there are multiple spans in a given input. they see the context, the context does not see them, inside the spans wrappers they are causal. each island can not see another island
4 maskq this token sees all the islands at the same time. the islands should not see it

This is the code I ran
!apt-get update -y && apt-get install -y software-properties-common 
!add-apt-repository ppa:git-core/ppa 
!apt-get install -y git 
!apt-get install -y build-essential 
%uv pip install torch ninja datasets matplotlib triton 
!rm -rf KernelDev
!git clone --branch remove-unnecessary-tasks https://github.com/Eupham/KernelDev.git

!python KernelDev/test_attention_behaviors.py
# !python KernelDev/entry.py --nproc_per_node=1 --config KernelDev/config.yaml --epochs 1 --precision bf16

These are the results
remote: Enumerating objects: 1794, done.
remote: Counting objects: 100% (485/485), done.
remote: Compressing objects: 100% (144/144), done.
remote: Total 1794 (delta 405), reused 364 (delta 341), pack-reused 1309 (from 4)
Receiving objects: 100% (1794/1794), 1.10 MiB | 13.75 MiB/s, done.
Resolving deltas: 100% (1173/1173), done.
============================================================
ATTENTION TOKEN BEHAVIOR TESTS
============================================================
Testing attention patterns for teacher forcing and cocktail party tasks
Note: These tests validate the attention logic without running Triton kernels

test_attention_mask_creation (__main__.AttentionBehaviorTests.test_attention_mask_creation)
Test creation of attention masks for different patterns. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Attention Mask Creation ===
Sequence length: 16
Is prefix: tensor([ True,  True,  True,  True, False, False, False, False, False, False,
        False, False, False, False, False, False])
In span: tensor([False, False, False, False, False, False, False, False,  True,  True,
         True,  True,  True,  True,  True, False])
Span ID: tensor([ 0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  2,  2,  2, -1])
Prefix positions: [0, 1, 2, 3]
Context positions: [4, 5, 6, 7]
Span 1 positions: [8, 9, 10, 11]
Span 2 positions: [12, 13, 14]
MASKQ positions: [15]
ok
test_attention_pattern_logic_validation (__main__.AttentionBehaviorTests.test_attention_pattern_logic_validation)
Test that the attention pattern logic itself is correctly implemented. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Attention Pattern Logic Validation ===
Test setup:
  is_prefix: tensor([ True,  True,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False])
  in_span: tensor([False, False, False, False, False,  True,  True,  True,  True,  True,
         True, False, False, False, False, False, False, False, False, False])
  span_id: tensor([ 0,  0,  0,  0,  0,  1,  1,  1,  2,  2,  2, -1,  0,  0,  0,  0,  0,  0,
         0,  0])

Validating attention pattern logic:
1. Prefix positions: [0, 1, 2]
   ✓ Prefix tokens should attend bidirectionally to each other
2. Context positions: [3, 4]
   ✓ Context tokens should attend causally + see prefix
3. Span 1 positions: [5, 6, 7]
   Span 2 positions: [8, 9, 10]
   ✓ Spans should be bidirectional within span, see context, not see each other
4. MASKQ positions: [11]
   ✓ MASKQ should see all spans and prefix
ok
test_cocktail_party_attention_patterns (__main__.AttentionBehaviorTests.test_cocktail_party_attention_patterns)
Test cocktail party attention behaviors. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Cocktail Party Attention Patterns ===
Test sequence shape: torch.Size([2, 32])
Prefix mask: tensor([ True,  True,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
In span mask: tensor([False, False, False, False, False, False, False,  True,  True,  True,
         True,  True,  True,  True,  True,  True, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
Span IDs: tensor([ 0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2, -1,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0])
Validation Results:
  ✓ prefix_bidirectional_batch_0: True
  ✓ context_causal_batch_0: True
  ✓ span_isolation_batch_0: True
  ✓ maskq_visibility_batch_0: True
  ✓ prefix_bidirectional_batch_1: True
  ✓ context_causal_batch_1: True
  ✓ span_isolation_batch_1: True
  ✓ maskq_visibility_batch_1: True
✓ All cocktail party attention patterns validated successfully!
ok
test_data_builder_cocktail_party_format (__main__.AttentionBehaviorTests.test_data_builder_cocktail_party_format)
Test data builder creates proper cocktail party format. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Data Builder Cocktail Party Format ===
Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 5 samples from the dataset.
Loading dataset: allenai/c4/en
Attempting Method 1: Load C4 'en' (streaming)...
README.md: 41.1kB [00:00, 96.0MB/s]
Resolving data files: 100%|████████████████| 1024/1024 [00:03<00:00, 268.20it/s]
Resolving data files: 100%|█████████████| 1024/1024 [00:00<00:00, 487621.17it/s]
C4 'en' (streaming) load_dataset call succeeded. Processing samples...
Reached max_samples (5) for C4 'en' streaming.
Finished processing C4 'en' streaming. Total valid samples extracted: 5
Successfully processed 5 samples from C4 'en' stream.
Successfully loaded 5 samples using C4 'en'.
Final dataset loaded with 5 samples.
Returning dataset splits: train=4, val=1, test=1
Tokenizing train split...
Text length for train: 3519 characters
Tokenized to 3519 byte tokens
Tokenizing validation split...
Text length for validation: 748 characters
Tokenized to 748 byte tokens
Tokenizing test split...
Text length for test: 1014 characters
Tokenized to 1016 byte tokens
train dataset: 3455 samples
Limited validation to 129 tokens for faster evaluation (target: 129)
validation dataset: 65 samples
Limited test to 129 tokens for faster evaluation (target: 129)
test dataset: 65 samples
Cocktail party dataloader created successfully
   Batch inputs shape: torch.Size([8, 64])
   Correct indices: tensor([2, 2, 1, 1, 0, 2, 2, 0])
   Contains CLS: False
   Contains SPAN: True
   Contains ES: True
   Contains MASKQ: True
   Sample text (first 32 tokens): x # backlinks per day I should d
   ⚠ CLS token not found (may be using fallback data without task prefixes)
✓ Data builder cocktail party format validated successfully!
ok
test_kernel_attention_patterns_cocktail_party (__main__.AttentionBehaviorTests.test_kernel_attention_patterns_cocktail_party)
Test that the actual kernel's attention patterns match expected behavior for cocktail party. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Kernel Attention Patterns - Cocktail Party ===
Test sequence shape: torch.Size([2, 32])
Prefix mask: tensor([ True,  True,  True, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
In span mask: tensor([False, False, False, False, False, False, False,  True,  True,  True,
         True,  True,  True,  True,  True,  True, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
Span IDs: tensor([ 0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  2,  2,  2,  2, -1,  0,
         0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0])
Created test tensors on cpu
Input tensor shapes: q=torch.Size([2, 4, 32, 16]), k=torch.Size([2, 4, 32, 16]), v=torch.Size([2, 4, 32, 16])
⚠ Kernel execution failed as expected on CPU (requires CUDA)
✓ API signature correctly accepts cocktail party metadata parameters
✓ Kernel API integration validated for cocktail party patterns!
ok
test_kernel_attention_patterns_teacher_forcing (__main__.AttentionBehaviorTests.test_kernel_attention_patterns_teacher_forcing)
Test that the actual kernel's attention patterns match expected behavior for teacher forcing. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Kernel Attention Patterns - Teacher Forcing ===
Test sequence shape: torch.Size([2, 32])
CLS position: 3
Prefix mask: tensor([ True,  True,  True,  True, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
Created test tensors on cpu
Input tensor shapes: q=torch.Size([2, 4, 32, 16]), k=torch.Size([2, 4, 32, 16]), v=torch.Size([2, 4, 32, 16])
⚠ Kernel execution failed as expected on CPU (requires CUDA)
✓ API signature correctly accepts return_attention_mask parameter
✓ Kernel API integration validated for teacher forcing patterns!
ok
test_kernel_mask_output_integration (__main__.AttentionBehaviorTests.test_kernel_mask_output_integration)
Test that demonstrates the complete integration of kernel mask output functionality. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Kernel Mask Output Integration ===
This test validates the complete solution to issue #177:
1. ✓ Modified flash_attention to accept return_attention_mask parameter
2. ✓ Updated kernel signature to support attention mask output
3. ✓ Added logic to write computed masks to output tensor
4. ✓ Made functionality toggleable via optional argument
5. ✓ Maintained existing kernel behavior when not requested

Test 1: Default behavior (no mask output)
✓ Expected CUDA requirement confirmed

Test 2: Mask output behavior
✓ Expected CUDA requirement confirmed for mask output

Test 3: Parameter validation
✓ All required parameters can be passed to flash_attention
✓ Cocktail party metadata integrated into kernel API
✓ return_attention_mask parameter controls output format

Test 4: Kernel modifications summary
✓ Added OUTPUT_ATTN_MASK tensor parameter to Triton kernel
✓ Added mask writing logic in attention computation loop
✓ Added RETURN_ATTENTION_MASK compile-time constant
✓ Kernel writes computed mask when requested
✓ No performance impact when mask output not requested

✓ Integration test completed successfully!
  Note: Actual CUDA execution would demonstrate mask correctness
  This test validates the complete API and integration changes
ok
test_special_token_behaviors (__main__.AttentionBehaviorTests.test_special_token_behaviors)
Test special token handling behaviors. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Special Token Behaviors ===
1. Testing PAD token handling...
   Sequence: tensor([10, 11,  1, 12,  0,  0,  0,  0])
   Attention mask: tensor([ True,  True,  True,  True, False, False, False, False])
   Valid length: 4
2. Testing CLS token detection...
   CLS token found at position: 2
3. Testing MASK token in context...
   MASK token found at position: 3
   MASK token should behave like regular context token (causal)
✓ All special token behaviors validated successfully!
ok
test_teacher_forcing_attention_patterns (__main__.AttentionBehaviorTests.test_teacher_forcing_attention_patterns)
Test teacher forcing attention behaviors. ... Using UTF-8 byte tokenization with vocabulary size: 262
Max evaluation tokens per split: 50000
Will attempt to load up to 10 samples from the dataset.

=== Testing Teacher Forcing Attention Patterns ===
Test sequence shape: torch.Size([2, 32])
CLS position: 3
Prefix mask: tensor([ True,  True,  True,  True, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False, False,
        False, False])
Validation Results:
  ✓ prefix_bidirectional_batch_0: True
  ✓ context_causal_batch_0: True
  ✓ cls_visibility_batch_0: True
  ✓ prefix_bidirectional_batch_1: True
  ✓ context_causal_batch_1: True
  ✓ cls_visibility_batch_1: True
✓ All teacher forcing attention patterns validated successfully!
ok

----------------------------------------------------------------------
Ran 9 tests in 7.774s

OK

============================================================
✓ ALL TESTS PASSED!
Attention token behaviors are correctly implemented.
============================================================
